### PR TITLE
Eigenpooling

### PR DESCRIPTION
**Pull request by thazhemadam**
Starting to sketch out EigenPooling layers which would resolve (at least a part of #10).

This will be largely based on the following work - [Graph Convolutional Networks with EigenPooling](https://arxiv.org/abs/1904.13107)

Presently, `AGNPool` simply does a max/mean pooling. The goal of this PR is to implement pooling layers that don't flatten the node representations into the graph representation.

This will be performed using a new type of pooling layer (named `EigenPool` for now).
TLDR of what these pooling layers would do -

- [ ] The initial graph adjacency matrix that represents the structure of the crystal would get coarsened into subgraphs.
- [ ] The node features of each "super-node" generated from coarsening is computed using the EigenPooling pooling operator.
- [ ] Finally, it returns a (new? or mutated?) `FeaturizedAtoms` object, that corresponds to the sub-graph.

### CODE REVIEW

File: src/layers/pool/eigenpool.jl [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    result = Vector()
+
+    for i = 1:size(L_eigen_vectors)[1]
+        L_i = L_eigen_vectors[:, i]
+        push!(result, L_i'*features)
+    end
+
+    # using an agreeable H and then return H elements of result hcatt-ed into a single 1xdH vector
+    result = hcat(result...)'
+    reshape(result, length(result), 1)  # return it as a dHx1 Matrix
```
Comment (rkurchin) [inline src/layers/pool/eigenpool.jl:38]: Yeah, but returning something of length `N x D` isn't going to work because ultimately the size can't depend on `N` if you want to be able to feed in graphs of different sizes.

As per my comment above, I think a sensible way to do this could be that there's a parameter H that says how many eigenvectors to keep, then we can guarantee a return length of `H x D`.

OR, if we're worried about that having inconsistent performance across graph sizes, we could instead specify a parameter (say `h`) in (0,1) that says the _fraction_ of eigenvectors to keep and then do a standard (e.g. max or mean) pooling across a list of `h` vectors of length `D` to always return a vector of length `D`.

File: src/layers/pool/eigenpool.jl [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+Graph Convolutional Networks with EigenPooling - Yao Ma, Suhang Wang, Charu C. Aggarwal, Jiliang Tang
+https://arxiv.org/abs/1904.13107
+
+Since the size of the matrices we are dealing with isn't as huge as the one in the original work,
+let us try using this as a global pooling mechanism instead, i.e., the adjacency matrix of the crystal
+itself would be only "subgraph" from which we "coarsen" to a single node and pool accordingly, which
+in theory would give us the overall graph representation
+=#
+
+# TBD - what other fields would be necessary for the pooling layer itself?
```
[the remark this answers is not in the document]
Comment (rkurchin) [reply src/layers/pool/eigenpool.jl:17]: (see more detailed comment below)

File: src/layers/layers.jl [hunk 3]
```diff
@@ -0,0 +1,48 @@
+module Layers
```
Comment (rkurchin) [inline src/layers/layers.jl:1]: I'm not sure this really needs to be a separate module since it's kind of the main/only thing the package does apart from the convenience functions for building standard model architectures, and I don't really see a risk of any sort of namespace conflicts...

File: src/layers/pool/eigenpool.jl (hunk 1 above)
Comment (thazhemadam) [reply src/layers/pool/eigenpool.jl:38]: > Yeah, but returning something of length N x D isn't going to work because ultimately the size can't depend on N if you want to be able to feed in graphs of different sizes.

Fair enough. We need to standardize across different graph sizes.

> As per my comment above, I think a sensible way to do this could be that there's a parameter H that says how many eigenvectors to keep, then we can guarantee a return length of H x D.

Would having `H::Integer` as a field in `EigenPool` be the best solution for this? I figured that `H` for a graph of size `N1` might not be appropriate (or rather, the "best") `H` for another graph of size `N2`.

Instead, what if, like `AGNPool`, we let users determine what they want their pooled feature length to be? Essentially,
`N * H` + `length(zero padding)` = `pooled_feature_length`

That way, pooled features would all be of similar lengths regardless of the graph sizes?

Comment (thazhemadam) [reply src/layers/pool/eigenpool.jl:48]: Addressed in https://github.com/Chemellia/AtomicGraphNets.jl/pull/90/commits/a38fdf43a9aee58a8f2fd9d172efcbd9d329111d
